components/containers/partitioned_vector/include/hpx/components/containers/partitioned_vector/partitioned_vector_segmented_iterator.hpp

```diff
@@ -1230,8 +1234,21 @@ namespace hpx { namespace traits {
     };
 
     template <typename T, typename Data>
-    struct is_value_proxy<hpx::segmented::detail::vector_value_proxy<T, Data>>
+    struct proxy_value<
+        hpx::segmented::detail::local_vector_value_proxy<T, Data>>
       : std::true_type
     {
     };
+
+    template <typename T, typename Data>
+    struct is_value_proxy<hpx::segmented::detail::vector_value_proxy<T, Data>>
+    {
+        using type = T;
+    };
+
+    template <typename T, typename Data>
+    struct proxy_value<hpx::segmented::detail::vector_value_proxy<T, Data>>
+    {
+        using type = T;
+    };
```

Comment:
Am I misunderstanding these traits or should this be:
```suggestion
    struct proxy_value<hpx::segmented::detail::local_vector_value_proxy<T, Data>>
    {
        using type = T;
    };

    template <typename T, typename Data>
    struct is_value_proxy<hpx::segmented::detail::vector_value_proxy<T, Data>>
      : std::true_type
    {
    };

    template <typename T, typename Data>
    struct proxy_value<hpx::segmented::detail::vector_value_proxy<T, Data>>
    {
        using type = T;
    };
```
?

Or should `is_value_proxy` simply be `true_type` for `T` and `Data` if and only if `proxy_value<T, Data>::type` exists?

Reply:
`is_value_proxy<T>` checks whether `T` is a type that is used as a proxy. `proxy_value<T>` allows to deduce the type that is returned by a proxy (`is_value_proxy<T>` is true). If `is_value_proxy<T>`  evaluates to false, then `proxy_value<T>` returns `T`.

Reply:
All right, that sounds like what I imagined. But then my first question still stands. Currently `struct proxy_value< hpx::segmented::detail::local_vector_value_proxy<T, Data>>` inherits from `true_type` when I would expect it to have a nested typedef `using type = T`. Likewise, `struct is_value_proxy<hpx::segmented::detail::vector_value_proxy<T, Data>>` has a nested typedef `using type = T` when I would expect it to inherit from `true_type`. In other words, I think there's been some sort mistaken transposition...

Reply:
I'm confused. The idea is for `is_value_proxy<T>` to be a boolean trait usable in an enable_if (or similar). For this reason it is derived from `true_type` or `false_type` depending on whether `T` is a proxy type or not. Note that `true_type` and `false_type` have an embedded `using type = ...;` referring to themselves.

The type `proxy_value<T>` is not meant to be used as a boolean trait, rather to extract the type returned by the proxy. It 'returns' the type that is provided by the proxy (if `is_value_proxy<T>` would return true). For usability reasons, it defaults to simply 'returning' `T` for non-proxy types (if `is_value_proxy<T>` would return false).

Reply:
Uhh, I think I see now what you're referring to. Yes - good catch, I messed it up. Will fix.

Reply:
That's fixed now. Thanks again for catching the issue.